src/siwe.rs

```diff
@@ -0,0 +1,357 @@
+        write!(f, "0x{}", hex::encode(&self.0))
+    }
+}
+
+impl FromStr for SIWESignature {
+    type Err = anyhow::Error;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s.split_once("0x") {
+            Some(("", h)) => Ok(Self(<[u8; 65]>::from_hex(h)?)),
+            _ => Err(anyhow!("Invalid hex string, no leading 0x")),
+        }
+    }
+}
+
+#[serde_as]
+#[derive(Serialize, Deserialize)]
+pub struct SIWEMessage(
+    #[serde_as(as = "DisplayFromStr")] Message,
+    #[serde_as(as = "DisplayFromStr")] SIWESignature,
+);
+
+impl Deref for SIWEMessage {
+    type Target = Message;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+pub struct SIWEZcapTokens {
+    pub invocation: KeplerInvocation,
+    pub delegation: SIWEMessage,
+}
+
+pub struct SIWETokens {
+    pub invocation: SIWEMessage,
+    pub delegation: Option<SIWEMessage>,
+    // kinda weird
+    pub orbit: Cid,
+    pub invoked_action: Action,
+}
+
+#[rocket::async_trait]
+impl<'r> FromRequest<'r> for SIWEZcapTokens {
+    type Error = anyhow::Error;
+    async fn from_request(request: &'r Request<'_>) -> Outcome<Self, Self::Error> {
+        match (
+            request.headers().get_one("x-kepler-invocation").map(|b64| {
+                base64::decode_config(b64, base64::URL_SAFE)
+                    .map_err(|e| anyhow!(e))
+                    .and_then(|s| serde_json::from_slice(&s).map_err(|e| anyhow!(e)))
+            }),
+            request.headers().get_one("x-siwe-delegation").map(|b64| {
+                base64::decode_config(b64, base64::URL_SAFE)
+                    .map_err(|e| anyhow!(e))
+                    .and_then(|s| serde_json::from_slice(&s).map_err(|e| anyhow!(e)))
+            }),
+        ) {
+            (Some(Ok(invocation)), Some(Ok(delegation))) => Outcome::Success(Self {
+                invocation,
+                delegation,
+            }),
+            (Some(Err(e)), _) => {
+                tracing::debug!("{}", e);
+                Outcome::Failure((Status::Unauthorized, e))
+            }
+            (_, Some(Err(e))) => {
+                tracing::debug!("{}", e);
+                Outcome::Failure((Status::Unauthorized, e))
+            }
+            (_, _) => Outcome::Forward(()),
+        }
+    }
+}
+
+#[rocket::async_trait]
+impl<'r> FromRequest<'r> for SIWETokens {
+    type Error = anyhow::Error;
+    async fn from_request(request: &'r Request<'_>) -> Outcome<Self, Self::Error> {
+        let (invocation, delegation): (SIWEMessage, Option<SIWEMessage>) = match (
+            request.headers().get_one("x-siwe-invocation").map(|b64| {
+                base64::decode_config(b64, base64::URL_SAFE)
+                    .map_err(|e| anyhow!(e))
+                    .and_then(|s| serde_json::from_slice(&s).map_err(|e| anyhow!(e)))
+            }),
+            request
+                .headers()
+                .get_one("x-siwe-delegation")
```

Comment:
I know we talked about it and said we would fix it later, but I'm going to put my initial idea: the Invocation and Delegation types (used in `ZCAPTokens`) should implement `FromRequest` themselves and that way you have any combination of invocation and delegation out of the box.

Reply:
this is a good suggestion, I've [created an issue to track it](https://github.com/spruceid/kepler/issues/81)